### PR TITLE
feat: add error boundaries and suspense usage with TRPC

### PR DESCRIPTION
**Pull request by austinwoon**
### Problem 1
Our current method of fetching data for pages with query params utilize `getServerSideProps`. We noticed a trend of slow initial load times for these pages as long as ~7s. This is likely (but not confirmed) attributed to Chakra's huge bundle size which slows down the time needed to build the server rendered page.

### Solutions considered to root problem

We had two options

- Attempt to fix `getServerSideProps`'s bundle issues with Chakra - possibly looking into how we can treeshake chakra for server rendered pages to reduce build times
- Introduce CSR patterns in our application

We went with option 2 as that was already in the existing roadmap.

### Problem 2

Existing CSR patterns with page based routing was cumbersome as we could not use React 18's `Suspense` and the default `ErrorBoundary` from `react-error-boundary`. 

This led to 2 dx issues:
- In every component where we use query, we had to track `isLoading` and `isError` states and render the fallback components accordingly
- There was a lack of a top level `ErrorBoundary` which would catch any errors thrown within child components which `useSuspenseQuery`

### Solution 
- The inability to use React's `Suspense` component was due to the fact that any child component which calls the `useRouter` hook erroring out during server side rendering as the router had not mounted. T
    - To mitigate this, we created our own `Suspense` wrapper that keeps track of a `isMounted` state. 
        -  If the router is ready, we will render React's `Suspense` with the children
        - If router is not ready, `fallbackComponent` will render.  
- Created our own `ErrorBoundary` wrapper which catches error boundary states to render the correct fallback components

**Improvements**:
- Refactored all queries to utilise `useSuspenseQuery`
- Updated logic to handle redirects between `sign-in` and `dashboard`
- Updated centralised logic for unauthorised errors redirecting to `sign-in`. This logic for this is now located in the `ErrorBoundary` component whenever a TRPCError has a code of `UNAUTHORIZED`

## Tests

- [x] Navigating to `dashboard` without logging in succesfully redirects me to `sign-in`
- [x] Navigating to `sign-in` whilst authed redirects me to either `callbackUrl` or `dashboard`
- [x] `ErrorBoundary` fallback component renders on different TRPCError codes
- [ ] `Suspense` fallback renders when data is suspending

### CODE REVIEW

File: src/components/ErrorBoundary/ErrorBoundary.tsx [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+      if (!res.success) return <UnexpectedErrorCard />
+
+      // The choice to not redirect via next's router was intentional to handle ErrorBoundary for the app root
+      // Using next's router.push('/sign-in') will not render the SignIn component as it won't be mounted in the app root as the ErrorBoundary fallback component will be rendered instead
+      // Using vanilla location redirecting will prompt a full page reload of /sign-in page, which will never trigger the root ErrorBoundary, thus rendering the full component correctly
+      if (res.data === 'UNAUTHORIZED') {
+        window.location.href = 'sign-in'
+        return
+      }
```
Comment (austinwoon) [inline src/components/ErrorBoundary/ErrorBoundary.tsx:49]: @karrui @dvrylc check out this statement for the root level ErrorBoundary and lmk what you think. We can walk through tomorrow if docs arent clear.

File: src/pages/_app.tsx [hunk 2]
```diff
@@ -9,26 +9,49 @@ import { DefaultLayout } from '~/templates/layouts/DefaultLayout'
...
 
-const MyApp = (({ Component, pageProps }: AppPropsWithAuthAndLayout) => {
-  const getLayout =
-    Component.getLayout ?? ((page) => <DefaultLayout>{page}</DefaultLayout>)
-
+const MyApp = (({
+  Component,
+  pageProps,
+  ...props
+}: AppPropsWithAuthAndLayout) => {
```
Comment (dvrylc) [inline src/pages/_app.tsx:24]: nit: no need to destructure the props, can just passthrough to `ChildWithLayout`

File: src/components/Suspense/Suspense.tsx [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    // isReady conditional is needed so that suspenseQueries do not fire twice.
+    // Without this, the child component will fire a query with an `undefined` query param followed by the actual query with param defined
+    if (router.isReady) {
+      setIsMounted(true)
+    }
+  }, [router.isReady])
+
+  if (isMounted) {
+    return <ReactSuspense {...props} />
+  }
```
Comment (dvrylc) [inline src/components/Suspense/Suspense.tsx:26]: actually do we really need the `isMounted` local state? thinking can just check `router.isReady` here

Comment (austinwoon) [reply src/components/Suspense/Suspense.tsx:26]: I tried before, but we get another error 

image

File: src/components/ErrorBoundary/ErrorBoundary.tsx [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    // Define a state variable to track whether is an error or not
+    this.state = { hasError: false }
+  }
+  static getDerivedStateFromError(error: unknown) {
+    // Update state so the next render will show the fallback UI
+    return { hasError: true, error }
+  }
+  componentDidCatch(error: Error) {
+    // You can use your own error logging service here
+    console.log({ error })
```
Comment (karrui) [inline src/components/ErrorBoundary/ErrorBoundary.tsx:23]: nit: prefer `console.error` if we are even outputting the error.  Might want to change the shape too and just pass in the error as-is:

```ts
console.error(error)
```

File: src/features/feedback/components/TeamFeedbackList.tsx [hunk 5]
```diff
@@ -1,19 +1,14 @@
 import { Box, Stack, StackDivider } from '@chakra-ui/react'
 import { useFilterFeedback } from '../api/useFilterFeedback'
-import { TeamFeedbackListSkeleton } from './TeamFeedbackListSkeleton'
 import { TeamFeedbackRow } from './TeamFeedbackRow'
 
 export const TeamFeedbackList = (): JSX.Element => {
-  const { filteredFeedback, isFetching } = useFilterFeedback()
-
-  if (!filteredFeedback) {
-    return <TeamFeedbackListSkeleton />
```
Comment (karrui) [inline src/features/feedback/components/TeamFeedbackList.tsx:10]: should we use this skeleton as a suspense fallback instead of removing it completely?

File: src/features/profile/api/useUser.ts [hunk 6]
```diff
@@ -1,27 +1,15 @@
...
-      ((!redirectIfFound && !user) ||
-        // If redirectIfFound is also set, redirect if the user was found
-        (redirectIfFound && user))
-
-    if (shouldRedirect) {
-      Router.push(redirectTo)
-    }
-  }, [user, redirectIfFound, redirectTo, isFetching])
+  const [user] = trpc.me.get.useSuspenseQuery(undefined, {
+    onSuccess: () => {
```
Comment (karrui) [inline src/features/profile/api/useUser.ts:7]: comment: do we no longer redirect if the user is unauthorised? seems like nowhere else in the code does the auth gate happen

File: src/pages/sign-in.tsx [hunk 7]
```diff
@@ -57,7 +58,7 @@ const SignIn = () => {
 
 export const getServerSideProps = withSessionSsr(
```
Comment (karrui) [inline src/pages/sign-in.tsx:59]: are we still using SSR here? I would have thought the intention was to migrate to CSR fully.

File: src/utils/trpc.ts [hunk 8]
```diff
@@ -79,14 +90,13 @@ export const trpc = createTRPCNext<AppRouter, SSRContext>({
...
             retry: (failureCount, error) => {
               if (error instanceof TRPCClientError) {
-                if (error.data?.code === 'FORBIDDEN') {
-                  Router.push('/forbidden')
-                }
+                const res = TRPCWithErrorCodeSchema.safeParse(error)
 
-                if (error.data?.code === 'UNAUTHORIZED') {
-                  Router.push('/sign-in')
+                if (res.success && NON_RETRYABLE_ERROR_CODES.has(res.data)) {
```
Comment (karrui) [inline src/utils/trpc.ts:98]: seems like a regression here, should we redirect to the appropriate pages if needed?

**Review by karrui — CHANGES_REQUESTED** [review]

File: src/features/feedback/components/TeamFeedbackList.tsx (hunk 5 above)
Comment (austinwoon) [reply src/features/feedback/components/TeamFeedbackList.tsx:10]: Skeleton is still used, please refer to `DashboardContainer` wrapper component within `Dashboard.tsx`

File: src/features/profile/api/useUser.ts (hunk 6 above)
Comment (austinwoon) [reply src/features/profile/api/useUser.ts:7]: We do, this redirected is done on the top level `ErrorBoundary` instead. I've wrapped the app root with an `ErrorBoundary`, and all `unauthorized` TRPCErrors will do a manual relocation. Code block below.

```     
 if (res.data === 'UNAUTHORIZED') {
        const params = new URLSearchParams(window.location.search)

        const callbackUrl = params.get('callbackUrl')

        window.location.href = !!callbackUrl
          ? `sign-in/?${CALLBACK_URL_KEY}=${callbackUrl}`
          : `sign-in`

        return
      }
```

File: src/pages/sign-in.tsx (hunk 7 above)
Comment (austinwoon) [reply src/pages/sign-in.tsx:59]: Will look to do some experimentation with CSR for auth later. I forgot what the exact blocker was in the sea of blockers i was dealing with. I think it had something to do with infinite redirects but it should've been solved with my other refactors

File: src/utils/trpc.ts (hunk 8 above)
Comment (austinwoon) [reply src/utils/trpc.ts:98]: Using `Router.push` here will not work if an error was thrown and the top level error boundary catches the error.

Instead of relying on router.push, im thinking we can set a convention to do all redirects in the `ErrorBoundary` component instead.

It's a little abstract, but we can walk through it later and I can demo what i mean.

to reproduce the behaviour, remove the unauthorized relocation block from `ErrorBoundary`, and try to redirect to `sign-in` from the `trpc` singleton. You'll see that the redirection does not load the `Sign-In` page as the component has not been mounted.

This is an edge case that only happens if the top level `ErrorBoundary` catches something and does not render the page with layout component

**Comment by karrui** [comment]
e2e tests failing, but fixed in #94. Can rebase/merge develop in first?

**Comment by austinwoon** [comment]
> e2e tests failing, but fixed in #94. Can rebase/merge develop in first?

done

**Review by karrui — APPROVED** [review]
lgtm
